src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+                    if needed.contains(&(idx + input_arity)) {
+                        col_map.insert(idx + input_arity, input_proj.len() + new_scalars.len());
+                        expr.visit_mut(&mut |e| {
+                            if let ScalarExpr::Column(c) = e {
+                                *c = *col_map.get(c).unwrap();
+                            }
+                        });
+                        new_scalars.push(expr.clone());
+                    }
+                }
+                for c in projection.iter_mut() {
+                    *c = *col_map.get(c).unwrap();
+                }
+                self.action(input, input_proj);
+                *relation = input.take_dangerous().map(new_scalars).project(projection);
+            }
+            RelationExpr::Join {
+                inputs,
+                equivalences,
+                ..
+            } => {
+                let input_mapper = expr::JoinInputMapper::new(inputs);
+                let mut needed: HashSet<_> = projection.iter().cloned().collect();
+                for equivalence in equivalences.iter() {
+                    for expr in equivalence {
+                        needed.extend(expr.support());
+                    }
+                }
+                let mut col_map = HashMap::new();
+
+                let mut total_arity = 0;
+
+                for (idx, input) in inputs.iter_mut().enumerate() {
+                    let mut proj = vec![];
+                    for col in 0..input.arity() {
+                        let global_col = input_mapper.map_column_to_global(col, idx);
+                        if needed.contains(&global_col) {
+                            col_map.insert(global_col, total_arity);
+                            total_arity += 1;
+                            proj.push(col);
+                        }
+                    }
+                    self.action(input, proj);
+                }
+
+                for equivs in equivalences.iter_mut() {
+                    for expr in equivs.iter_mut() {
+                        expr.visit_mut(&mut |e| {
+                            if let ScalarExpr::Column(c) = e {
+                                *c = *col_map.get(c).unwrap();
+                            }
+                        });
+                    }
+                }
+
+                *relation = RelationExpr::join_scalars(inputs.to_vec(), equivalences.clone())
+                    .project(
+                        projection
+                            .into_iter()
+                            .map(|c| *col_map.get(&c).unwrap())
+                            .collect(),
+                    );
+            }
+            RelationExpr::Reduce {
+                input,
+                group_key,
+                aggregates,
+                monotonic: _,
+            } => {
+                // TODO(justin); do we need to preserve `monotonic` here?
```

Comment:
It might get refreshed later, but it should just stay put anyhow (it is a property of `input` that has been pushed to this operator for rendering information).